File: app/connections.py
"""Manage remote S3 connections."""
from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Dict, List, Optional

from .config import AppConfig


@dataclass
class RemoteConnection:
    id: str
    name: str
    endpoint_url: str
    access_key: str
    secret_key: str
    region: str = "us-east-1"


class ConnectionStore:
    def __init__(self, config_path: Path) -> None:
        self.config_path = config_path
        self._connections: Dict[str, RemoteConnection] = {}
        self.reload()

    def reload(self) -> None:
        if not self.config_path.exists():
            self._connections = {}
            return

        try:
            with open(self.config_path, "r") as f:
                data = json.load(f)
                for item in data:
                    conn = RemoteConnection(**item)
                    self._connections[conn.id] = conn
        except (OSError, json.JSONDecodeError):
            self._connections = {}

    def save(self) -> None:
        self.config_path.parent.mkdir(parents=True, exist_ok=True)
        data = [asdict(conn) for conn in self._connections.values()]
        with open(self.config_path, "w") as f:
            json.dump(data, f, indent=2)

    def list(self) -> List[RemoteConnection]:
        return list(self._connections.values())

    def get(self, connection_id: str) -> Optional[RemoteConnection]:
        return self._connections.get(connection_id)

    def add(self, connection: RemoteConnection) -> None:
        self._connections[connection.id] = connection
        self.save()

    def delete(self, connection_id: str) -> None:
        if connection_id in self._connections:
            del self._connections[connection_id]
            self.save()

File: rust/myfsio-engine/crates/myfsio-server/src/stores/connections.rs
use std::path::{Path, PathBuf};
use std::sync::Arc;

use parking_lot::RwLock;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RemoteConnection {
    pub id: String,
    pub name: String,
    pub endpoint_url: String,
    pub access_key: String,
    pub secret_key: String,
    #[serde(default = "default_region")]
    pub region: String,
}

fn default_region() -> String {
    "us-east-1".to_string()
}

pub struct ConnectionStore {
    path: PathBuf,
    inner: Arc<RwLock<Vec<RemoteConnection>>>,
}

impl ConnectionStore {
    pub fn new(storage_root: &Path) -> Self {
        let path = storage_root
            .join(".myfsio.sys")
            .join("config")
            .join("connections.json");
        let inner = Arc::new(RwLock::new(load_from_disk(&path)));
        Self { path, inner }
    }

    pub fn reload(&self) {
        let loaded = load_from_disk(&self.path);
        *self.inner.write() = loaded;
    }

    pub fn list(&self) -> Vec<RemoteConnection> {
        self.inner.read().clone()
    }

    pub fn get(&self, id: &str) -> Option<RemoteConnection> {
        self.inner.read().iter().find(|c| c.id == id).cloned()
    }

    pub fn add(&self, connection: RemoteConnection) -> std::io::Result<()> {
        {
            let mut guard = self.inner.write();
            if let Some(existing) = guard.iter_mut().find(|c| c.id == connection.id) {
                *existing = connection;
            } else {
                guard.push(connection);
            }
        }
        self.save()
    }

    pub fn delete(&self, id: &str) -> std::io::Result<bool> {
        let removed = {
            let mut guard = self.inner.write();
            let before = guard.len();
            guard.retain(|c| c.id != id);
            guard.len() != before
        };
        if removed {
            self.save()?;
        }
        Ok(removed)
    }

    fn save(&self) -> std::io::Result<()> {
        if let Some(parent) = self.path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let snapshot = self.inner.read().clone();
        let bytes = serde_json::to_vec_pretty(&snapshot)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
        std::fs::write(&self.path, bytes)
    }
}

fn load_from_disk(path: &Path) -> Vec<RemoteConnection> {
    if !path.exists() {
        return Vec::new();
    }
    match std::fs::read_to_string(path) {
        Ok(text) => serde_json::from_str(&text).unwrap_or_default(),
        Err(_) => Vec::new(),
    }
}
